flang/lib/Optimizer/Transforms/SimplifyFIROperations.cpp

```diff
@@ -162,7 +173,52 @@ class DoConcurrentConversion
     assert(loop.getRegion().hasOneBlock());
     mlir::Block &loopBlock = loop.getRegion().getBlocks().front();
 
-    // Collect iteration variable(s) allocations do that we can move them
+    // Handle localization
+    if (!loop.getLocalVars().empty()) {
+      mlir::OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(&loop.getRegion().front());
+
+      std::optional<mlir::ArrayAttr> localSyms = loop.getLocalSyms();
+
+      for (auto [localVar, localArg, localizerSym] : llvm::zip_equal(
+               loop.getLocalVars(), loop.getRegionLocalArgs(), *localSyms)) {
+        mlir::SymbolRefAttr localizerName =
+            llvm::cast<mlir::SymbolRefAttr>(localizerSym);
+        fir::LocalitySpecifierOp localizer = findLocalizer(loop, localizerName);
+
+        mlir::Value localAlloc =
+            rewriter.create<fir::AllocaOp>(loop.getLoc(), localizer.getType());
```

Comment:
I wonder if this should be done on the heap because it will happen for every loop iteration. Does the spec say that it has to be on the stack?

Reply:
Nothing in the spec says it has to be on the stack AFAICT. However, this is following the behavior of `handleLocalitySpecs` in `Bridge.cpp` (without delayed localization). Added a todo, let me know if you disagree with looking into this later.

Reply:
Please could you use TODO(loc, "message") so that the compiler crashes (in a controlled way) instead of producing incorrect code if any of these do sneak through.

Reply:
I think you meant for the `init` and `dealloc` todo right?

Reply:
All good. Done.